packages/react-async/src/context.ts

```diff
@@ -0,0 +1,23 @@
+import { AsyncOptions, KEYS_OF_FETCHOPTIONS } from "./types"
+
+/**
+ * Removes all "known "keys from fetchOptions leaving only the
+ * keys that are not defined in the type AsyncOptions.
+ *
+ * One special behavior is when it encounters a `context` object
+ * inside of the `options`. It will treat them as defaults for
+ * the context object. But they are overwritten by props from
+ * the options.
+ */
+export function createContext<T, C>(options: AsyncOptions<T, C>): C {
+  const context: AsyncOptions<T, C> = {
+    ...options.context,
+    ...options,
+  }
+
+  KEYS_OF_FETCHOPTIONS.forEach(k => {
+    delete context[k]
+  })
```

Comment:
It seems this means it is impossible to have keys in context that conflict with `AsyncOptions` ... 

Shouldn't it delete entries from `options` but not from `options.context`  ?

Given that this PR is breaking change maybe it should only support passing arguments through `options.context` to bypass this processing ?

Reply:
+1 for only passing arguments through `options.context`. It'll be a pretty big change but for good reason. Still having the option to also use rest props on `AsyncProps` only complicates things.

I would also like to see the `context` arg on `run` to be merged on top of `options.context`.

Reply:
@ghengeveld so just to be sure you want the api to change from:

```jsx
<Async promiseFn={loadPlayer} playerId={1} suspense={true}>

useAsync(loadPlayer,  { playerId: 1, suspense: true});

// Alt
useAsync({ promiseFn: loadPlayer,  playerId: 1, suspense: true});
```

To:

```jsx
<Async promiseFn={loadPlayer} context={{ playerId: 1}} suspense={true}>

useAsync(loadPlayer, { context: { playerId: 1}, suspense: true});

// Alt
useAsync({ promiseFn: loadPlayer, context: { playerId: 1}, suspense: true});
```

I think this would be a great change.

I'm also for making the `context` pure, as in nothing gets magically added to it or deleted from it. This will make the code more readable.

But I'd also recommend overloading `useAsync` in the future to allow:

```ts
useAsync(loadPlayer,  { playerId: 1}, {suspense: true});
```

As I think this would result in the most readable api.

Reply:
If we are changing APIs here anyways I'm all for moving from positional parameters to named parameters (one parameter with a keyed object).

So I'd prefer to be only left with 
```ts
<Async promiseFn={loadPlayer} context={{ playerId: 1}} suspense={true}>

useAsync({ promiseFn: loadPlayer, context: { playerId: 1}, suspense: true});
```
as that will make it more readable at a glance, allow for easier future expansion and will keep the typings more readable/less error-prone.

Reply:
I'm in favor of dropping the overloaded signatures and only support the single object argument. The shorthand where you pass in the async function as first argument was added to support some very basic use cases where it'd be the _only_ argument, but it's not worth the extra complexity in code and API consistency.

Reply:
By the way we're going to have to write a codemod for this, because it would otherwise require quite a bit of manual work to upgrade.

Reply:
> I'm also for making the context pure, as in nothing gets magically added to it or deleted from it. This will make the code more readable.

Agreed. People can always spread it themselves.

Reply:
I've updated the code so:

1. There is no more overload of useAsync.
2. There is no more magical context spreading / building.